embassy-stm32f4/src/exti.rs

```diff
@@ -0,0 +1,128 @@
+//! Async low power Serial.
+//!
+//! The peripheral is autmatically enabled and disabled as required to save power.
+//! Lowest power consumption can only be guaranteed if the send receive futures
+//! are dropped correctly (e.g. not using `mem::forget()`).
+
+use core::future::Future;
+use core::mem;
+use core::pin::Pin;
+
+use embassy::gpio::{WaitForFallingEdge, WaitForRisingEdge};
+use embassy::interrupt::OwnedInterrupt;
+use embassy::util::Signal;
+
+use crate::hal::gpio::{Edge, ExtiPin as HalExtiPin};
+use crate::hal::prelude::*;
+use crate::hal::syscfg::SysCfg;
+use crate::pac::EXTI;
+
+use crate::interrupt;
+
+static mut exti: mem::MaybeUninit<EXTI> = mem::MaybeUninit::<EXTI>::uninit();
+
+
+pub struct ExtiContainer {}
+
+impl ExtiContainer {
+    pub fn new(ex: EXTI) -> Self {
+        unsafe { exti.write(ex); }
+
+        Self{}
+    }
+}
+
+
+/// Interface to the Serial peripheral
+pub struct ExtiPin<T: HalExtiPin, I: OwnedInterrupt> {
+    pin: T,
+    signal: Signal<()>,
+    interrupt: I,
+}
+
+/*
+    Irq	Handler	Description
+    EXTI0_IRQn	EXTI0_IRQHandler	Handler for pins connected to line 0
+    EXTI1_IRQn	EXTI1_IRQHandler	Handler for pins connected to line 1
+    EXTI2_IRQn	EXTI2_IRQHandler	Handler for pins connected to line 2
+    EXTI3_IRQn	EXTI3_IRQHandler	Handler for pins connected to line 3
+    EXTI4_IRQn	EXTI4_IRQHandler	Handler for pins connected to line 4
+    EXTI9_5_IRQn	EXTI9_5_IRQHandler	Handler for pins connected to line 5 to 9
+    EXTI15_10_IRQn	EXTI15_10_IRQHandler	Handler for pins connected to line 10 to 15
+*/
+
+impl<T: HalExtiPin, I: OwnedInterrupt> ExtiPin<T, I> {
+    pub fn new(mut pin: T, container: &ExtiContainer, syscfg: &mut SysCfg, interrupt: I) -> Self {
```

Comment:
Lifetime of the ExtiPin should be linked to the lifetime of the ExtiContainer. Otherwise you could create the ExtiContainer, create some ExtiPins, then `free()` the ExtiContainer (if it was implemented) , then keep using the pins.

Pins shouldn't be usable if the ExtiContainer has been freed, because that means the user has got the `&mut EXTI` back, and can now manually touch the EXTI registers which could break things.